sympy/printing/printer.py

```diff
@@ -347,39 +347,39 @@ def _as_ordered_terms(self, expr, order=None):
             return expr.as_ordered_terms(order=order)
 
 
-class _PrintFunction:
-    """
-    Function wrapper to replace ``**settings`` in the signature with printer defaults
-    """
-    def __init__(self, f, print_cls: Type[Printer]):
-        # find all the non-setting arguments
-        params = list(inspect.signature(f).parameters.values())
-        assert params.pop(-1).kind == inspect.Parameter.VAR_KEYWORD
-        self.__other_params = params
-
-        self.__print_cls = print_cls
-        update_wrapper(self, f)
-
-    def __repr__(self) -> str:
-        return repr(self.__wrapped__)  # type:ignore
-
-    def __call__(self, *args, **kwargs):
-        return self.__wrapped__(*args, **kwargs)
-
-    @property
-    def __signature__(self) -> inspect.Signature:
-        settings = self.__print_cls._get_initial_settings()
-        return inspect.Signature(
-            parameters=self.__other_params + [
-                inspect.Parameter(k, inspect.Parameter.KEYWORD_ONLY, default=v)
-                for k, v in settings.items()
-            ],
-            return_annotation=self.__wrapped__.__annotations__.get('return', inspect.Signature.empty)  # type:ignore
-        )
-
-
 def print_function(print_cls):
     """ A decorator to replace kwargs with the printer settings in __signature__ """
-    def decorator(f):
-        return _PrintFunction(f, print_cls)
-    return decorator
```

Comment:
How about localizing the hack to right here?
```python
def decorator(f):
    # We have to create a subclass so that `help` actually shows the docstring. IPython and Sphinx do not
    # need this, only a raw python terminal.
    cls = type(f'{f.__qualname__}_doc_hack', (_PrintFunction,), dict(__doc__=f.__doc__))
    return cls(f, print_cls)
```

Reply:
As remarked in https://github.com/sympy/sympy/issues/20231#issuecomment-715897919, this is only necessary in Python 3.6, 3.7, and 3.8.

Reply:
That is probably fine. I haven't looked at if any of the other attributes updated in update_wrapper are broken from doing it on a class instance. It's good to see that Python 3.9 actually does the right thing here now. 

Actually, perhaps a bigger concern is that in both cases, we are doing this at import time, so it's not great in terms of keeping the import time down. Is there really no way to monkeypatch the `__signature__` of a function?

Reply:
I guess it's so that changing the global defaults changes the signature. I'm not really sure if this is worth all this metaprogramming.

Reply:
> Is there really no way to monkeypatch the __signature__ of a function?

Not if we want it to update to reflect the global defaults set by init_printing

> Actually, perhaps a bigger concern is that in both cases, we are doing this at import time, so it's not great in terms of keeping the import time down.

I don't think class creation is expensive, so neither patch should make a difference over master. The expensive bit is the sanity check in `__init__` that calls inspect.signature, which could be skipped.

Reply:
> I'm not really sure if this is worth all this metaprogramming.

I'd argue that the reuse across lots of printers makes this worthwhile.

Reply:
OK. It's really unfortunate that the `__signature__` API uses an attribute instead of a method call. I've run into this problem in other projects, where you have to do some metaprogramming to make `__signature__` give a dynamic thing (the good news is to do so on a class only requires using a descriptor, but that won't work for functions like `latex`).